### ISSUE TITLE
page.open work synchronously

### ISSUE THREAD

**Issue by oliseviche** [amir20/phantomjs-node#591]
### Expected behavior

Page.open must open pages in async mode.

### Actual behavior

Now pages are opened in sync mode. So we will wait for completion of all pages piece by piece.

### Steps to reproduce the behavior
```
phantom.create([], {
 phantomPath: './phantomjs'
}).then(instance => {
 return Promise.all([instance.createPage(), instance.createPage()]);
}).then(pages => {
 pages[0].open(#long_running_url).then(() => console.log('long runing end'));
 pages[1].open(#fast_running_url).then(() => console.log('fast runing end'));
});
```
Would log 'long runing end' and only after 'fast runing end', but expected opposite.

### Environment

- OS: Ubuntu 14.04
- Node version (`node -v`): 7.2.0
- package version: ^3.0.1

**Comment by amir20** [comment]
I don't think your "expected" behavior is valid. PhantomJS is single threaded. If you were to write this in vanilla PhantomJs, it would only handle one page at a time. 

If you do require both pages to work in parallel then why don't you just create two phantom instances? 

Making as "upstream" dependency. Please close if this is answers your question.

**Comment by dbauszus-glx** [comment]
I am actually pulling my hair out about a similar issue. I have an application which creates multi page reports. A page takes about 20 seconds to complete since I need to wait for resources to be loaded. If two users create a report at the same time. i.e. the page content for the second user is set before the page is rendered for the first user then either the process is closed ph.exit() when the page for the first user is rendered. Or the second user gets the page from the first user if I comment the ph.exit().

```
function pdf_export_map(site, siteData, req, _pdfExport_census) {
        phantom.create()
            .then(instance => {
                ph = instance;
                return instance.createPage();
            })
            .then(p => {
                page = p;
                page.open('about:blank');
            })
            .then(function () {
                var renderTimeout,
                    c = 0;

                function doRender() {
                    console.log('page loaded');
                    var d = Date.now();
                    var filename = 'm' + d.toString() + '.pdf';
                    page.render('tmp/' + filename)
                        .then(function () {
                            ph.exit();
                            req.fileStack.push(process.env.FILEPATH + filename);
                            _pdfExport_census(site, siteData, req)
                        });
                }

                page.on('onResourceRequested', function (r) {
                    c++;
                    console.log(('000' + r.id).slice(-4) + ' | ' + r.url);
                    clearTimeout(renderTimeout);
                });

                page.on('onResourceReceived', function (r) {
                    if (!r.stage || r.stage === 'end') {
                        c--;
                        console.log(('000' + r.id).slice(-4) + ' | ' + r.status);
                        if (c === 0) {
                            renderTimeout = setTimeout(doRender, 5000);
                        }
                    }
                });

                if (process.platform == 'win32') {
                    page.property('paperSize', {
                        width: '11in',
                        height: '8.5in'
                    });
                } else {
                    page.property('paperSize', {
                        width: '1056px',
                        height: '816px'
                    });
                }

                page.property('viewportSize', {
                    width: 1056,
                    height: 816
                });

                page.property('clipRect', {
                    top: 0,
                    left: 0,
                    width: 1056,
                    height: 816
                });

                ejs.renderFile('views/phantommap.ejs', {
                    site: site,
                    features: siteData.features,
                    subdirectory: process.env.SUBDIRECTORY,
                    page: req.fileStack.length + 1
                }, function (err, html) {
                    page.setContent(html, 'http://' + process.env.HOST + '/amazon/phantommap');
                });

            })
            .catch(err => {
                console.log(err);
                ph.exit();
            });
}

```

I just posted this on stackoverflow. No idea how to get this to play ball and have one page created after the next.

**Comment by amir20** [comment]
I responded to the stackoverflow question.

**Comment by oliseviche** [comment]
Hi amir20!

I've tested something like this in vanilla phantomJS:
/*---------------test.js------------------------
const page = require("page");
page.open("long_run", () => { log("long"); });
page.open("fast_run", () => { log("fast"); });
/*----------------------------------------------
and run it like "phantomjs test.js". The output was in async fashion - it log fast and long.

I've also tried the solution with new instance of phantom for each request, but unfortunately, the stress tests show that 100 request in 100 seconds consumes all available memory on testing machine, even when the page and instance are closed and exited after finish. I assume that it could also happen because of application, which I am open with phantom.

May be there are some good practices exists how to deliver phantom semi-to high load usage experience?

**Comment by amir20** [comment]
There is. You should use a pool like https://github.com/amir20/phantomjs-node/issues/590.

How were you able to run ` () => { log("long"); }` in phantom? `() => {}` syntax is not supported. At least, I thought it isn't. Now I am doubting.

**Comment by amir20** [comment]
Also I found https://groups.google.com/forum/#!topic/phantomjs/ytehrJKxu6M/discussion which suggests it is single like I had thought. Maybe something has changed since 2015.

**Comment by oliseviche** [comment]
I wrote arrow function just for briefness here, it not a real copy-paste. But anyway, thank you for your attention.

**Comment by amir20** [comment]
Now I am curious if this is actually bug. I'll try to reproduce it. I'll reopen if I think it's something I can fix. It might not be fixable due to the way process communication works with PhantomJs.

**Comment by elias-winberg** [comment]
I can confirm that this is a "bug". `system.stdin.readline` is blocking in PhantomJS, which means that the next command isn't read/processed until the previous one has finished.

If you add
```js
console.log('Reading...');
```
at the very start of the `read` function in `shim/index.js`, and execute the following code
```js
const Phantom = require('phantom');

let phantom;

Phantom.create().then(function (instance) {
    phantom = instance;
    return Promise.all([phantom.createPage(), phantom.createPage()]);
}).then(function (pages) {
    pages[0].open('https://www.google.com').then(console.log);
    pages[1].open('https://www.google.com').then(console.log);
}).then(function () {
    phantom.exit();
});
```
you'll see:
```bash
...
info: Reading...
success
info: Reading...
success
info: Reading...
...
```

**Comment by amir20** [comment]
Hmm. good point! Thanks. Do you think there is a fix we can do @elias-winberg ?

**Comment by elias-winberg** [comment]
@amir20 maybe we could listen for commands using a [Web Server](http://phantomjs.org/api/webserver/) instead?

**Comment by amir20** [comment]
That's what it used to do. It caused memory leaks, security exceptions and a nightmare to run in cluster mode. :/ I had to find random ports to open so that I can run them with a pool of instances. 

But good idea!

**Comment by elias-winberg** [comment]
> I had to find random ports to open so that I can run them with a pool of instances.

Yea, that's a major issue. Hmm, any other ideas?

**Comment by amir20** [comment]
There is a heartbeat right now implemented that send `NOOP` requests to phantomjs every 100ms. That should have made `system.stdin.readline` not block. But I think if two requests are milleseconds away from each other then it doesn't help much. 

I'll have to keep thinking about it, but for now I will reopen this ticket for a short while. If we don't find a solution because phantomjs's limitations then I will close it.

**Comment by KpjComp** [comment]
The solution to this is a `pool`,  I'm using one and it can process about 24 PDF's in about 5 seconds using a pool of 3 phantomJs's.. 

`return Promise.all([phantom.createPage(), phantom.createPage()]);`

That is not a good idea, your trying to send 2 commands at the same time,. Don't try and do more than 2 things inside a phantomJS, it's going to give you nothing but trouble.  The IPC message loop would just get out of sync.  Phantomjs-node is using a simple protocol of Request/Response,  So Request/Request/Response will not work, the protocol would need updating with messageID's to cope with out of sync requests.

The whole idea of the `NOOP` was to help with implementing a pool,. So that idle Pool's would take very little CPU.  If you want I could show my Pool implementation, or maybe we could implement a Pool into phantomjs-node.

**Comment by amir20** [comment]
@KpjComp thanks for the response! Although one thing might not be true.

> Phantomjs-node is using a simple protocol of Request/Response, So Request/Request/Response will not work, the protocol would need updating with messageID's to cope with out of sync requests.

You can actually send multiple requests. `request.id` is generated and tracks which request correlates to each response.  So technically it would all work. 

I did some investigation on this and was able to reproduce this ticket. Both worked fine, but I couldn't figure out why it wasn't doing both at the same time. The interesting thing is that I see `NOOP` being processed and then suddenly it pauses. It might be that while the page is rendering for the first one, nothing else can happen. 

There is a phantom pool link that was added to the bottom of the readme. I suggest using that.

**Comment by KpjComp** [comment]
Oh, should have looked at source code again before commenting. :)

Anyway, I think I might have spotted the issue.

`this.id = crypto.createHash('md5').update(target + name + params.join()).digest('hex');`

If you repeat a command with the same args, your going to get the same id.
Your `this.commands.delete(command.id);`  is now going to delete both commands.

Why not just create an increasing `id`, ?..

**Comment by amir20** [comment]
Interesting. I didnt not think about

> IOW: [instance.createPage(), instance.createPage()] both of those createPage requests are going to have the same id.. the last instance.creatPage resolve will get fired, but the first was overwritten so will never get resolved.

That might fix the bug. I used actually have autogenerated id. But it was slow. So replaced with hash. Let me try it again with this.

**Comment by amir20** [comment]
I was not able to figure this out. I made a couple of changes doing https://github.com/amir20/phantomjs-node/compare/sync-bug?expand=1. 

1. I made the heartbeat every 10ms not 100ms
2. I removed `if(command in progress)`
3. I made `command.id` a unique int via `++`
4. I added logs for `NOOP`

After compiling and doing 

```js
const Phantom = require('phantom');

let phantom;

Phantom.create().then(function(instance) {
    phantom = instance;
    return Promise.all([phantom.createPage(), phantom.createPage()]);
}).then(function(pages) {
    pages[0].open('http://localhost:3000/slow').then(function() {console.log('slow')});
    pages[1].open('http://localhost:3000/fast').then(function() {console.log('fast')});
}).then(function() {
    phantom.exit();
});
```

I still see
```
⨉⪧ ~/Workspace/phantomjs-node/examples on sync-bug ± ⟩ env DEBUG=true node bug.js
debug: Starting /usr/local/bin/phantomjs /Users/araminfar/Workspace/phantomjs-node/examples/node_modules/phantom/lib/shim/index.js
debug: Sending: {"id":1,"target":"phantom","name":"createPage","params":[],"deferred":null}
debug: Sending: {"id":2,"target":"phantom","name":"createPage","params":[],"deferred":null}
debug: Sending NOOP command.
debug: Parsing: {"id":1,"target":"phantom","name":"createPage","params":[],"deferred":null,"response":{"pageId":1}}
debug: Parsing: {"id":2,"target":"phantom","name":"createPage","params":[],"deferred":null,"response":{"pageId":2}}
debug: Sending: {"id":3,"target":"page$1","name":"invokeAsyncMethod","params":["open","http://localhost:3000/slow"],"deferred":null}
debug: Sending: {"id":4,"target":"page$2","name":"invokeAsyncMethod","params":["open","http://localhost:3000/fast"],"deferred":null}
debug: Sending: {"id":5,"target":"phantom","name":"invokeMethod","params":["exit"],"deferred":null}
debug: Received NOOP command.
debug: Parsing: {"id":3,"target":"page$1","name":"invokeAsyncMethod","params":["open","http://localhost:3000/slow"],"deferred":null,"response":"success"}
slow
debug: Parsing: {"id":4,"target":"page$2","name":"invokeAsyncMethod","params":["open","http://localhost:3000/fast"],"deferred":null,"response":"success"}
fast
debug: Parsing: {"id":5,"target":"phantom","name":"invokeMethod","params":["exit"],"deferred":null}
```

The id is being incremented correctly. NOOP for some reason is not being sent every 10ms. It almost looks like phantomjs is not processing. It all comes to `page.open` which is implemented as follows

```js
invokeAsyncMethod: function(command) {
        let target = objectSpace[command.target]; // this is a page object
        target[command.params[0]].apply(target, command.params.slice(1).concat(result => {
            command.response = result;
            completeCommand(command);
        })); // This adds a callback to last parameter 
    },
```

Unless the above is blocking (which I can't imagine it would be) then I don't understand why this is behaving. 

@KpjComp thoughts?

**Comment by KpjComp** [comment]
You test is wrong..

```
    pages[0].open('http://localhost:3000/slow').then(function() {console.log('slow')});
    pages[1].open('http://localhost:3000/fast').then(function() {console.log('fast')});
```

The above code is not waiting on the promise's, so your `exit` is getting fired too early.
Remember if you don't return from a thenable, your `then` method will end prematurely.

Change to the following..

```
return Promise.all([
    pages[0].open('http://localhost:3000/slow').then(function() {console.log('slow')}),
    pages[1].open('http://localhost:3000/fast').then(function() {console.log('fast')})
]);
```

**Comment by amir20** [comment]
I had actually done `Promise.all` and I noticed it didn't make a difference. Because the output is still `slow` then `fast`. 

```
debug: Starting /usr/local/bin/phantomjs /Users/araminfar/Workspace/phantomjs-node/examples/node_modules/phantom/lib/shim/index.js
debug: Sending: {"id":1,"target":"phantom","name":"createPage","params":[],"deferred":null}
debug: Sending: {"id":2,"target":"phantom","name":"createPage","params":[],"deferred":null}
debug: Sending NOOP command.
debug: Parsing: {"id":1,"target":"phantom","name":"createPage","params":[],"deferred":null,"response":{"pageId":1}}
debug: Parsing: {"id":2,"target":"phantom","name":"createPage","params":[],"deferred":null,"response":{"pageId":2}}
debug: Sending: {"id":3,"target":"page$1","name":"invokeAsyncMethod","params":["open","http://localhost:3000/slow"],"deferred":null}
debug: Sending: {"id":4,"target":"page$2","name":"invokeAsyncMethod","params":["open","http://localhost:3000/fast"],"deferred":null}
debug: Received NOOP command.
debug: Sending NOOP command.
debug: Parsing: {"id":3,"target":"page$1","name":"invokeAsyncMethod","params":["open","http://localhost:3000/slow"],"deferred":null,"response":"success"}
slow
debug: Parsing: {"id":4,"target":"page$2","name":"invokeAsyncMethod","params":["open","http://localhost:3000/fast"],"deferred":null,"response":"success"}
fast
debug: Sending: {"id":5,"target":"phantom","name":"invokeMethod","params":["exit"],"deferred":null}
debug: Received NOOP command.
debug: Parsing: {"id":5,"target":"phantom","name":"invokeMethod","params":["exit"],"deferred":null}
```

I had realized I was exiting too soon. But when I fixed it, it still didn't make a difference. I just don't get why nothing is happening while `slow` request is executing. 

In regards to your other suggestion. We already do reject all commands when exiting. 

```js
this.process.on('exit', code => {
            this.logger.debug(`Child exited with code {${code}}`);
            this._rejectAllCommands(`Phantom process stopped with exit code ${code}`);
```

Which does 

```js
 /**
     * rejects all commands in this.commands
     */
    _rejectAllCommands(errmsg: string = 'Phantom exited prematurely'): void {
        // prevent heartbeat from preventing this from terminating
        clearInterval(this.heartBeatId);
        for (const command of this.commands.values()) {
            if (command.deferred != null) {
                command.deferred.reject(new Error(errmsg));
            }
        }
    }
```

So maybe you missed that. It didn't make sense to wait for all the commads to be done. If someone calls `exit()` then really it should exit and reject all commands.

**Comment by KpjComp** [comment]
> Because the output is still slow then fast.

That's because we see if any commands are in transit, and if there is we don't NOOP, it was like that before too.

Because the slow will take time, they will be commands in transit.

Easy solution, remove the check for commands in transit.

eg. Below remove the `this.commands.size === 0` check..
```
_heartBeat(): void {
    if (this.commands.size === 0 && !this.isNoOpInProgress) {
        this.isNoOpInProgress = true;
        this.process.stdin.write(NOOP + os.EOL, 'utf8');
    }
}
```

**Comment by amir20** [comment]
I think you totally missed my comment [above](https://github.com/amir20/phantomjs-node/issues/591#issuecomment-265621173) 😆  I made a [branch](https://github.com/amir20/phantomjs-node/compare/sync-bug?expand=1) to fix this. I did actually remove `this.commands.size === 0`. 

You can try pulling down the branch and building that locally. I am honestly just stumped that phantomjs process just sits there doing nothing. For a minute, I thought your idea about `command.id` was going to fix it.

**Comment by KpjComp** [comment]
I'm not really sure now!!..

A couple of things to try,..  Firstly what is your fast/slow HTTP server code look like.

One theory I have, is does "phantom.open" block until it at least gets some data streaming.  It might be worth adding some extra logging inside the `invokeAsyncMethod:`  that state's when the command was called, and when it finished (add some timestamp to logging output).  PhantomJS might be blocking badly, to detect when page load is done.  It wound't surprise me if this was the case, the fact that readline does not have an async variant does not bode well here.

**Comment by amir20** [comment]
The server is at https://github.com/amir20/slow-server/blob/master/index.js

I haven't had time to do more logging. But that's next I guess.

**Comment by amir20** [comment]
> One theory I have, is does "phantom.open" block until it at least gets some data streaming

If that was true, then wouldn't vanilla phantom do the same? I have tried it and it does work in parallel correctly with just phantomjs.

**Comment by KpjComp** [comment]
Ok, I'm going to do an install and see this happening..  The only other thing that I can think of is that Stdio is blocking, but our NOOP's should be unblocking the readline, allowing the stdout to send data.

**Comment by KpjComp** [comment]
Ok figured it out,... :)

Were only calling ` setTimeout(read, 0);` on `function completeCommand(command) `

If you want messages to overlap, we cal call the read sooner..

eg.. In function read do ->

```
        transform(command.params);
        try {
            executeCommand(command);
            setTimeout(read, 0);  //if we want to process command's overlapping we need this here
        } catch (e) {
            command.error = e.message;
            completeCommand(command);
        }

```

And then we can remove ->

```
function completeCommand(command) {
    _system2.default.stdout.writeLine('>' + JSON.stringify(command));
    // Prevent event-queue from clogging up by reads that block.
    //    setTimeout(read, 0);  //don't need this here now,.  
}

```

Doing these changes I now get `fast running end`  `long running end`

**Comment by amir20** [comment]
Aw man! Thanks. Pull request coming soon! It makes sense.

**Comment by KpjComp** [comment]
No problem's.   

It actually got me thinking, now we know we can do multiple page requests in parallel, I'm wondering if a phantomjs pool is even required.  Javascript is single threaded, but I believe Chrome is multi-threaded per page.

So what might be useful is a Page Pool, rather than a Phantom Pool.  This might give even better performance as there might be more chance for phantomjs to cache resources, and an instance of Chrome can take a fair bit of memory.

It would still make sense though to recycle phantomJS's instances, just in case of memory leaks,.  Eg.  Say after 50 page requests or a certain amount of time do a recycle.  And to prevent recycle lag, it could be made so that the other instance fires up a few seconds beforehand.

**Comment by amir20** [comment]
New version released!

**Comment by amir20** [comment]
@oliseviche  this should be fixed now. Make sure to thank @KpjComp for his hard work. :)

**Comment by oliseviche** [comment]
Hi @amir20 and @KpjComp !

Great job! Thanks both of you for all hard work you've done! I will check it right away after the weekends.

**Comment by Attrash-Islam** [comment]
I wrote something based on this library that runs tasks in sync
https://github.com/Attrash-Islam/phantom-tasker (Not completed yet)

You don't need to apply harmony for this phantom-tasker since it compiled through typescript 2+ which support async/await syntax in ES3/ES5! :tada: :tada: :tada:
